orchagent/aclorch.cpp

```diff
@@ -493,17 +510,110 @@ bool AclRuleL3::validateAddAction(string attr_name, string _attr_value)
     }
+    else if (attr_value.find(PACKET_ACTION_REDIRECT) != string::npos)
+    {
+        // resize attr_value to remove argument, _attr_value still has the argument
+        attr_value.resize(string(PACKET_ACTION_REDIRECT).length());
+
+        sai_object_id_t param_id = getRedirectObjectId(_attr_value);
+        if (param_id == SAI_NULL_OBJECT_ID)
+        {
+            return false;
+        }
+        value.aclaction.parameter.oid = param_id;
+    }
     else
     {
         return false;
     }
     value.aclaction.enable = true;
 
-    m_actions[aclActionLookup[attr_name]] = value;
+    m_actions[aclL3ActionLookup[attr_value]] = value;
 
     return true;
 }
 
+// This method should return sai attribute id of the redirect destination
+sai_object_id_t AclRuleL3::getRedirectObjectId(const string& redirect_value)
+{
+    // check that we have a colon after redirect rule
+    size_t colon_pos = string(PACKET_ACTION_REDIRECT).length();
+    if (redirect_value[colon_pos] != ':')
+    {
+        SWSS_LOG_ERROR("Redirect action rule must have ':' after REDIRECT");
+        return SAI_NULL_OBJECT_ID;
+    }
+
+    if (colon_pos + 1 == redirect_value.length())
+    {
+        SWSS_LOG_ERROR("Redirect action rule must have a target after 'REDIRECT:' action");
+        return SAI_NULL_OBJECT_ID;
+    }
+
+    string target = redirect_value.substr(colon_pos + 1);
+
+    // Try to parse physical port and LAG first
+    Port port;
+    if(m_pAclOrch->m_portOrch->getPort(target, port))
+    {
+        if (port.m_type == Port::PHY)
+        {
+            return port.m_port_id;
+        }
+        else if (port.m_type == Port::LAG)
+        {
+            return port.m_lag_id;
+        }
+        else
+        {
+            SWSS_LOG_ERROR("Wrong port type for REDIRECT action. Only physical ports and LAG ports are supported");
+            return SAI_NULL_OBJECT_ID;
+        }
+    }
+
+    // Try to parse nexthop ip address
+    try
+    {
+        IpAddress ip(target);
+        if (!m_pAclOrch->m_neighOrch->hasNextHop(ip))
+        {
+            SWSS_LOG_ERROR("ACL Redirect action target next hop ip: '%s' doesn't exist on the switch", ip.to_string().c_str());
+            return SAI_NULL_OBJECT_ID;
+        }
+
+        m_redirect_target_next_hop = target;
+        m_pAclOrch->m_neighOrch->increaseNextHopRefCount(ip);
+        return m_pAclOrch->m_neighOrch->getNextHopId(ip);
+    }
+    catch (...)
+    {
+        // no error, just try next variant
+    }
+
+    // try to parse nh group ip addresses
+    try
+    {
+        IpAddresses ips(target);
+        if (!m_pAclOrch->m_routeOrch->hasNextHopGroup(ips))
+        {
+            SWSS_LOG_ERROR("ACL Redirect action target next hop ip list: '%s' doesn't exist on the switch", ips.to_string().c_str());
+            return SAI_NULL_OBJECT_ID;
+        }
+
+        m_redirect_target_next_hop_group = target;
+        m_pAclOrch->m_routeOrch->increaseNextHopRefCount(ips);
```

Comment:
If you increase the next hop ref count here, and failed at last to add this ACL rule, the ref count will not be decreased.